packages/react-native/android/src/main/java/com/bugsnag/reactnative/BugsnagReactNative.java

```diff
@@ -1,103 +1,193 @@
+        Client client = Bugsnag.init(context, apiKey);
+        configureClient(client);
+        return client;
+    }
+
+    /**
+     * Instantiates a bugsnag client with a given configuration object.
+     *
+     * @param context the application context
+     * @param config  configuration for how bugsnag should behave
+     * @return the bugsnag client
+     */
+    public static Client startWithConfiguration(Context context, Configuration config) {
+        Client client = Bugsnag.init(context, config);
+        configureClient(client);
+        return client;
+    }
+
+    private static void configureClient(Client client) {
+        InternalHooks.configureClient(client);
+
+        // TODO set these as config up front if possible
+        client.setIgnoreClasses("com.facebook.react.common.JavascriptException");
+        bugsnagAndroidVersion = client.getClass().getPackage().getSpecificationVersion();
+        // TODO set codeBundleId here
+        Log.d("BugsnagReactNative", "Initialised bugsnag-react-native");
+    }
+
     @Override
     public String getName() {
         return "BugsnagReactNative";
     }
 
-    /** Start a new session. */
+    @ReactMethod(isBlockingSynchronousMethod = true)
+    public void getConfig(Promise promise) {
+        Configuration config = Bugsnag.getClient().getConfig();
+        // TODO serialise as a readablemap
+    }
+
+    /**
+     * Updates the native metadata using metadata from the JS layer
+     */
     @ReactMethod
-    public void startSession() {
-        Bugsnag.startSession();
+    public void updateMetaData(ReadableMap update) {
+        // TODO update metadata
     }
 
-    /** Stop the current session. */
+    /**
+     * Updates the native context using the context from the JS layer
+     */
     @ReactMethod
-    public void stopSession() {
-        Bugsnag.stopSession();
+    public void updateContext(String context) {
+        // TODO update context
     }
 
-    /** Resume the previously started session or start a new one if none available. */
+    /**
+     * Updates the native user using the user from the JS layer
+     */
     @ReactMethod
-    public void resumeSession() {
-        Bugsnag.resumeSession();
+    void updateUser(String id, String name, String email) {
+        // TODO update user
     }
 
-    /** Leaves a breadcrumb. */
+    /**
+     * Deliver the report using the native delivery mechanism
+     */
     @ReactMethod
-    public void leaveBreadcrumb(ReadableMap options) {
-      // Bugsnag.leaveBreadcrumb(
-      //     options.getString("name"),
-      //     parseBreadcrumbType(options.getString("type")),
-      //     readStringMap(options.getMap("metadata")));
+    public void dispatch(ReadableMap payload, Promise promise) {
+        // TODO deserialize stacktrace etc here
+        BugsnagException exc = new BugsnagException("", "", new StackTraceElement[]{});
+
+        Bugsnag.getClient().notify(exc, new Callback() {
```

Comment:
`notify()` always blocks while it is capturing an error report, then delivery is typically async. If the promise doesn't resolve, that indicates a much bigger problem as the notifier will have crashed.

Reply:
Ok so I checked and `notify()`'s return value is `void`. I guess we should at least attempt to reject the promise when notify throws an error:

```java
try {
  Bugsnag.getClient().notify(exc, new Callback() {
  });
} catch (Exception e) {
  promise.reject("ERROR_CODE", e);
}
```

And unless there is a way to pass a callback to `notify()` which is run when the report is sent/saved to denote that it was sent/saved successfully, that'll have to be it. I guess if the notifier has crashed, the JS promise not ever being resolved/rejected is the least of the application's worries!

Reply:
If `notify` throws an exception then the process will die - I think we should avoid trying to catch that as it would indicate a bug with the Android notifier that probably isn't recoverable.